gen/generate.go

```diff
@@ -284,45 +292,69 @@ func generateModule(m *compile.Module, i thriftPackageImporter, builder *generat
 			return nil, err
 		}
 
-		buff := new(bytes.Buffer)
-		if err := g.Write(buff, nil /* fset */); err != nil {
-			return nil, fmt.Errorf(
-				"could not generate idl.go for %q: %v", m.ThriftPath, err)
-		}
+		if len(o.SingleFile) == 0 {
+			buff := new(bytes.Buffer)
+			if err := g.Write(buff, nil); err != nil {
+				return nil, fmt.Errorf(
+					"could not generate idl.go for %q: %v", m.ThriftPath, err)
+			}
 
-		files["idl.go"] = buff.Bytes()
+			files["idl.go"] = buff.Bytes()
+		}
 	}
 
 	// Services must be generated last because names of user-defined types take
 	// precedence over the names we pick for the service types.
 	if len(m.Services) > 0 {
-		for _, serviceName := range sortStringKeys(m.Services) {
-			service := m.Services[serviceName]
-
-			// generateModule gets called only for those modules for which we
-			// need to generate code. With --no-recurse, generateModule is
-			// called only on the root file specified by the user and not its
-			// included modules. Only services defined in these files are
-			// considered root services; plugins will generate code only for
-			// root services, even though they have information about the
-			// whole service tree.
-			if _, err := builder.AddRootService(service); err != nil {
-				return nil, err
-			}
-
-			serviceFiles, err := Service(g, service)
+		if len(o.SingleFile) > 0 {
+			serviceFiles, err := Services(g, m.Services, o.SingleFile)
 			if err != nil {
 				return nil, fmt.Errorf(
-					"could not generate code for service %q: %v",
-					serviceName, err)
+					"could not generate code for all services in the thrift file: %v", err)
 			}
 
 			if !o.NoServiceHelpers {
 				for name, buff := range serviceFiles {
 					files[name] = buff.Bytes()
 				}
 			}
+		} else {
+			for _, serviceName := range sortStringKeys(m.Services) {
+				service := m.Services[serviceName]
+
+				// generateModule gets called only for those modules for which we
+				// need to generate code. With --no-recurse, generateModule is
+				// called only on the root file specified by the user and not its
+
+				// considered root services; plugins will generate code only for
+				// root services, even though they have information about the
+				// whole service tree.
+				if _, err := builder.AddRootService(service); err != nil {
+					return nil, err
+				}
+
+				serviceFiles, err := Service(g, service)
+				if err != nil {
+					return nil, fmt.Errorf(
+						"could not generate code for service %q: %v",
+						serviceName, err)
+				}
+
+				if !o.NoServiceHelpers {
+					for name, buff := range serviceFiles {
+						files[name] = buff.Bytes()
+					}
+				}
+			}
 		}
+	} else if len(o.SingleFile) > 0 {
+		buff := new(bytes.Buffer)
+		if err := g.Write(buff, nil); err != nil {
+			return nil, fmt.Errorf(
+				"could not write non-service code when no services were found, but the single-service flag was specified: %v", err)
```

Comment:
I don't fully understand this error output. The "single-service flag" part doesn't make sense; from what I understand, the flag only specifies that everything gets written into a single file, but I see some error output about services, so it's confusing.

Reply:
Agreed, I've changed the error message. The original motivation behind the error message was to give context to my worry about the case in which the single-file flag was specified, but because there were no services defined in the thrift file no file would be written. In such as case, a file should always be written. I've now separated the logic to more naturally express that a file should be written if the single-file flag is specified, and therefore the error message can be changed to something more understandable.